docs/decisions/0007-lti-1p3-decoupling.rst

```diff
@@ -0,0 +1,115 @@
+1.3 launches from running outside of the platform.
+
+Open edX would like to support the following LTI use cases:
+
+* content launches in an XBlock context
+* content launches from multiple XBlocks, reusing LTIConfigurations between them (i.e. course context)
+* content launches outside an XBlock context in the platform (e.g. LTI course tabs, LTI 1.3 embedded discussions)
+* content launches from an independently deployable application (i.e. IDA or Django context)
+
+https://github.com/openedx/xblock-lti-consumer/pull/254 decoupled the LTI 1.3 launch handler endpoint
+``launch_gate_endpoint``, the LTI 1.3 keyset endpoint ``public_keyset_endpoint``, and the LTI 1.3 access token endpoint
+``access_token_endpoint`` from the XBlock by moving these endpoints into the Django plugin. This was the first step in
+decoupling the XBlock from the LTI 1.3 launch. This decision describes how to fully decouple the LTI 1.3 launch from the
+XBlock.
+
+Decision
+********
+
+* We will remove references to the ``location`` field from all views that implement the generic LTI 1.3 launch. This
+  includes the ``launch_gate_endpoint``, ``public_keyset_endpoint``, and ``access_token_endpoint`` views.
+* The responsibility of determining the launch information included in an LTI launch (e.g. roles and context claims)
+  will be that of the context in which the LTI 1.3 launch occurs (e.g. XBlock, IDA).
+* The library will expose a function that will return the URL that initiates the LTI 1.3 launch. The function will have
+  a parameter ``claims``, which will be an object that represents the claims that should be included in the LTI 1.3
+  launch.
+* Each context will supply the launch information to the LTI 1.3 launch by including it as an argument to the
+  aforementioned function. Each context will call this function to fetch the URL to trigger an LTI 1.3 launch.
+* In order for this launch information to be available throughout the LTI 1.3 launch, the library will include the
+  launch information provided by the context in the third-party initiated login request through the ``lti_message_hint``
+  query or form parameter.
+* The third-party initiated login request ``login_hint`` query or form parameter will be used to include the user
+  identifier. The ``login_hint`` will not be used to identify the ``LtiConfiguration`` instance. Information necessary
+  to identify the ``LtiConfiguration`` instance will be included in the ``lti_message_hint`` query or form parameter
+  instead.
+* We will not describe how to decouple LTI Advantage Services in this document. This may be done at a later time.
+
+Consequences
+************
+
+* LTI 1.3 launches will be decoupled from the XBlock and from the edX platform. This will allow Open edX to better
+  leverage LTI across the platform and across the ecosystem.
+* LTI Advantage Services will not be decoupled from the XBlock and from the edX platform. This means that LTI Advantage
```

Comment:
a side consequence we may want to mention is how this decision impacts the LTIConfiguration data model. Because we haven't tackled these advantage services yet location still needs to exist as a field right?

Reply:
I think we can find easy replacements for the location field for Deep Linking and Assignments and Grades services. 
These are the methods that can potentially break LTI Advantage when removing the the reference:
1. https://github.com/openedx/xblock-lti-consumer/blob/master/lti_consumer/signals.py#L16 - Links grades back to LMS (we should probably split this into (1) a signal emitted by the pure LTI function and (2) a receiver on the block side to react and link the grade).
2. Permissions and authentication for REST API endpoints: https://github.com/openedx/xblock-lti-consumer/tree/0a1f26234f9515524d117f5032ad3d8ff2bfb871/lti_consumer/lti_1p3/extensions/rest_framework
3. Declarative LTI AGS mode: https://github.com/openedx/xblock-lti-consumer/blob/0a1f26234f9515524d117f5032ad3d8ff2bfb871/lti_consumer/models.py#L420
4. This `UsageKey` reference: https://github.com/openedx/xblock-lti-consumer/blob/0a1f26234f9515524d117f5032ad3d8ff2bfb871/lti_consumer/models.py#L596

Reply:
@Zacharis278 I'm a bit late to reply, so I'm not sure this answer will be up-to-date. Feel free to ignore if so.

>Because we haven't tackled these advantage services yet location still needs to exist as a field right?

IIRC Not necessarily - as long as tools know which context they are being launched to, they should be able to send back unique identifiers. This only affects LTI advantage services, since they can store data that is placement specific.

E.g.: **AGS** stores grades for each placement, so it needs to know what is the location for each placement (in the Xblock launch case).
**DL** configures different placements to have different contents show up - so it needs to store a unique placement ID somewhere.

Reply:
_Note:_ about my previous comment with the code links - those are just places I know to be breaking points, no need to do anything related to them here in the ADR, but good to link to who will be implementing the code changes.